extensions/pdh/PDHCounters.h

```diff
@@ -0,0 +1,98 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <TCHAR.h>
+#include <pdh.h>
+#include <pdhmsg.h>
+#include <string>
+
+#include "PerformanceDataCounter.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+class PDHCounterBase : public PerformanceDataCounter {
+ public:
+  virtual ~PDHCounterBase() {}
+  static PDHCounterBase* createPDHCounter(const std::string& query_name, bool is_double = true);
+
+  const std::string& getName() const;
+  virtual std::string getObjectName() const;
+  virtual std::string getCounterName() const;
+  virtual PDH_STATUS addToQuery(PDH_HQUERY& pdh_query) = 0;
+  virtual PDH_STATUS collectData() = 0;
+ protected:
+  PDHCounterBase(const std::string& query_name, bool is_double) : PerformanceDataCounter(),
+    is_double_format_(is_double), pdh_english_counter_name_(query_name) {
+  }
+  DWORD getDWFormat() const;
+  const bool is_double_format_;
+  std::string pdh_english_counter_name_;
+};
+
+class PDHCounter : public PDHCounterBase {
+ public:
+  void addToJson(rapidjson::Value& body, rapidjson::Document::AllocatorType& alloc) const override;
+
+  PDH_STATUS addToQuery(PDH_HQUERY& pdh_query) override;
+  PDH_STATUS collectData() override;
+
+ protected:
+  friend PDHCounterBase* PDHCounterBase::createPDHCounter(const std::string& query_name, bool is_double);
```

Comment:
What's the reason for being this strict with the instantiation of the PDHCounters? Why is the instantiation restricted to the factory function only?

Reply:
Yeah, the pdh counter naming convention is strict so I wanted to prevent these classes to be created with invalid names.
This makes accessor functions to be clean (without checking boilerplate)